hal/src/nRF52840/i2c_hal.cpp

```diff
@@ -235,31 +239,61 @@ static int twi_init(HAL_I2C_Interface i2c) {
     return 0;
 }
 
-void HAL_I2C_Init(HAL_I2C_Interface i2c, void* reserved) {
+static bool HAL_I2C_Config_Is_Valid(const HAL_I2C_Config* config) {
+    if ((config == nullptr) ||
+            (config->rx_buffer == nullptr ||
+             config->rx_buffer_size == 0 ||
+             config->tx_buffer == nullptr ||
+             config->tx_buffer_size == 0)) {
+        return false;
+    }
+
+    return true;
+}
+
+int HAL_I2C_Init(HAL_I2C_Interface i2c, const HAL_I2C_Config* config) {
+    // Disable threading to create the I2C mutex
     os_thread_scheduling(false, NULL);
     if (m_i2c_map[i2c].mutex == NULL) {
         os_mutex_recursive_create(&m_i2c_map[i2c].mutex);
     } else {
         // Already initialized
         os_thread_scheduling(true, NULL);
-        return;
+        return SYSTEM_ERROR_NONE;
     }
 
+    // Capture the mutex and re-enable threading
     HAL_I2C_Acquire(i2c, NULL);
     os_thread_scheduling(true, NULL);
 
+    // Initialize internal data structure
+    if (HAL_I2C_Config_Is_Valid(config)) {
+        m_i2c_map[i2c].rx_buf = config->rx_buffer;
```

Comment:
This buffer can potentially be freed by user application. Same to `config->tx_buffer`.

Reply:
What do you propose we do? It's not supposed to be freed.

Reply:
We should allocate the buffer in HAL and return error if failed to do that, instead of inheriting the pointer to the buffer that is allocated in user application. We cannot guarantee that user don't free the buffer.

Reply:
Besides, the `config->rx_buffer` can be a temporary buffer allocated in stack in user application. In that case, the buffer is freed automatically. Unless we document it with bold font.

Reply:
> We should allocate the buffer in HAL and return error if failed to do that, instead of inheriting the pointer to the buffer that is allocated in user application. 

I disagree. We follow the same API style we already have for increasing the buffer size for Serial: https://docs.particle.io/reference/device-os/firmware/electron/#acquireserialbuffer-

Reply:
This also has the additional benefit of not allocating these buffers if the application does not use I2C at all.

Reply:
Btw, USART buffers are also provided by the application: https://github.com/particle-iot/device-os/blob/develop/wiring/src/spark_wiring_usartserial.cpp#L134

Reply:
We can allocate the buffer in `HAL_I2C_Init()` and free it in `HAL_I2C_Reset()`. That wouldn't consume RAM if I2C is not used. It's fine to leave it as-is, and just keep in mind that there is still risk to mess up the RAM data if the buffer is free unexpectedly.

Reply:
Let's leave it as-is and just make sure that the documentation clearly states that these buffers need to be preserved.

The difference between allocating completely in HAL vs in application, is that the application may provide statically allocated buffers, which somewhat controls the potential heap allocation errors. Additionally custom schemes may be used where a buffer is shared between multiple peripherals as long as they are not used concurrently.